### DISCUSSION TITLE
How can I integrate pydantic v2 URLs in code?

### DISCUSSION THREAD

**Discussion by bunny-therapist** [pydantic/pydantic#6395] [Pydantic V2]
In pydantic v1, I could set a field to be "AnyHttpUrl", and the format would be validated. The field value would be of type AnyHttpUrl, but that itself is ultimately a str. However, in pydantic v2, instead the field value would be of type "pydantic_core._pydantic_core.Url'" which is not a str. In other words, the following would print "True" in pydantic v1 but "False" in pydantic v2:
```py
import pydantic

class MyModel(pydantic.BaseModel):
    url: pydantic.AnyHttpUrl

x = MyModel(url="http://google.com")

print(isinstance(x, str))

```
In pydantic v1, the url is a str, which means it has all the methods of a str and passes all isinstance checks of a str. In pydantic v2, it is of a type which is an internal pydantic class. I have therefore no idea how to integrate this in my code. Consider the following example code:

```py
import pydantic
import requests

class MyModel(pydantic.BaseModel):
    url: pydantic.AnyHttpUrl


def get_from_url(url: str) -> requests.Response:
    return requests.get(url)

x = MyModel(url="http://google.com")
response = get_from_url(url=x.url)

```
In pydantic v1, this would just work. I have function which takes a str, so I pass in x.url to it. However, in pydantic v2, x.url is NOT a str anymore, it is actually of a type which is internal to pydantic. Therefore it does not seem feasible to use it in type hints, and I cannot really use its methods (I mean, it is a private class, I should not have to know its methods - my type checker/IDE certainly does not know). It seems like the only thing I can do is to convert x.url to str - `str(x.url)` - before using it - *every single time I access the field*.

How are issues like this supposed to be resolved? Am I supposed to never access fields without converting them?

(I ran into this issue when trying to go from pydantic v1 to v2, and I got lots of errors saying, e.g., `AttributeError: 'pydantic_core._pydantic_core.Url' object has no attribute 'startswith'` and similar. My code is obviously designed to handle the URLs as strings.)

**Comment by g0di** [comment]
I was about to start the exact same conversation before bumping on yours. The thing is, I'm wondering if this was even expected because I could not find any mention of this change neither in the doc nor the migration guide nor the future plan.

I'm supporting the fact that having URLs inherit from `str` was really useful and IMO made totally sense (because an URL is ultimately a string). Moreover, most tools/libraries dealing with URLs expect and use `str` in place of URLs so this force to have to wrap all our pydantic `Url` with `str(my_url)`.

This is not actually a big deal to me but I'm really wondering the reason behind (if any) and also to have this mentionned somewhere

**Comment by bunny-therapist** [comment]
A small correction: since pydantic.AnyHttpUrl has a defined python stub interface, type hinting using it actually works.

However, I am still stuck either calling str(...) or .unicode_string() on it everywhere (not sure which one of the two is best-practice, or if there is a difference between them). It is certainly less practical than when they were strings.

**Comment by mzamini92** [comment]
pydantic.AnyHttpUrl field has a defined Python stub interface but you still need to convert it to a str before you can use it in most other ways. If you just need to convert the field to a string that you can pass to another function, then the str() function is probably the best choice. If you need to access the underlying Unicode string representation of the URL, then the .unicode_string() method is the best choice. Here is an example of how you can use the str() function to convert an AnyHttpUrl field to a str:
```
import pydantic

class MyModel(pydantic.BaseModel):
    url: pydantic.AnyHttpUrl

x = MyModel(url="http://google.com")

print(str(x.url)) # This will print "http://google.com"
```

**Comment by phdesign** [comment]
Using an [Annotated Validator](https://docs.pydantic.dev/latest/concepts/validators/#annotated-validators) to convert the serialised format to str seems to work for me. I'm new to pydantic though so there may be some downsides to this approach that I'm not seeing.

```py
from typing_extensions import Annotated
from pydantic import AfterValidator, BaseModel, HttpUrl

HttpUrlString = Annotated[HttpUrl, AfterValidator(lambda v: str(v))]


class MyModel(BaseModel):
    url: HttpUrlString


x = MyModel(url="http://google.com")

print(x.url)
print(isinstance(x.url, str))

# Output:
# http://google.com/
# True
```

**Reply by antont** [reply]
Thanks - found this by googling the prob, am using this solution now too, while curious about how this is meant to be actually.

**Reply by morcef** [reply]
Not sure this is the case for all IDEs, but PyCharm still sees `url` as `Url | Url` in this case. So warnings will be raised when trying to pass it to a function that expects a string.

![image](https://github.com/pydantic/pydantic/assets/15701746/7f294203-0879-4508-addc-e6c5e88bd860)

**Reply by benoit-bnp** [reply]
I tried this solution (with FastAPI, for an input parameter of a route) but I get the following warning when testing: ```Expected `url` but got `str` - serialized value may not be as expected```. If anyone has an idea to prevent that, I'd gladly take it.

**Reply by awwitecki** [reply]
One of solutions could be declaring a field as `str` and validate it manually though for sure it's not necessarily elegant:
```py
from typing import Annotated

from pydantic import BaseModel, BeforeValidator, HttpUrl, TypeAdapter

http_url_adapter = TypeAdapter(HttpUrl)

Url = Annotated[str, BeforeValidator(lambda value: str(http_url_adapter.validate_python(value)))]


class Model(BaseModel):
    url: Url
```

EDIT: Adapter moved out from Annotated as suggested

**Reply by offlinehacker** [reply]
This is great, why should URL data format be anything but string after all.

**Reply by samuelcolvin** [reply]
You should move the type adapter out of the validator for performance.

**Comment by ramsrib** [comment]
Ran into [this](https://github.com/pydantic/pydantic/issues/7186) bug (which causes the url to be appended with "/") after fixing converting the url to str.

```py
url = "http://google.com"
assert str(HttpUrl(url)) == url # it fails with Pydantic 2.0 - 2.5.3
```

**Comment by kfreezen** [comment]
This is one of the many things that is causing me grief in a move from pydantic v1 to pydantic v2.

Now, I love pydantic v2, and its speed increases, but running into stuff like this makes me wish I hadn't moved my project to it. Making changes like these are fine for my code, but then when I have to require other developers to also make these little changes to their code, it causes them heartache and pain.

@samuelcolvin if you would only give us a way to have the old behavior back, I would be eternally grateful. Or at least grateful for the next 70 years.

If I could import something like the following to enable the behavior, it would ease our migration path. Or even a way to implement the old behavior with an equivalent class that we use in our own code. It doesn't matter.

```python
from pydantic.legacy import AnyHttpUrl
```

Thank you for reading.

**Reply by antont** [reply]
Have you tried the

```python
HttpUrlString = Annotated[HttpUrl, AfterValidator(lambda v: str(v))]
```

thing from above? I'm using it, mostly happily IIRC

I'm not saying that the lib should not provide it, but just wondering if that first aid bandage works for you or not .. as am not sure how well it works for me.

**Reply by samuelcolvin** [reply]
Or indeed simple

```py
HttpUrlString = Annotated[HttpUrl, AfterValidator(str)]
```

**Reply by antont** [reply]
> Or indeed simple
> 
> 
> 
> ```py
> 
> HttpUrlString = Annotated[HttpUrl, AfterValidator(str)]
> 
> ```

Right, I actually saw the lambda version in our codebase recently and was wondering why the do-nothing lambda was there .. but didn't dare to touch it. Thanks!

**Reply by lupodellasleppa** [reply]
@kfreezen  `pydantic.v1` actually exists, but if you were to try and use it for a single schema, then you'd have to use the `BaseModel` from `pydantic.v1` as well for that schema, as well as all other pydantic types from `pydantic.v1`, as the pydantic types and the `BaseModel` from `pydantic` v2 are not compatible. This could cause a lot of headache having to remember to use, for example, `model_dump()` instead of `dict()`, or `model_validate()` instead of `from_orm()`, only for that particular model everyhwere it is used.

Also, this workaround
```python
HttpUrlString = Annotated[HttpUrl, AfterValidator(str)]
```
produces the warning described in the [comment above](https://github.com/pydantic/pydantic/discussions/6395#discussioncomment-7356952)

The [provided solution](https://github.com/pydantic/pydantic/discussions/6395#discussioncomment-7361416) to this causes **mypy** errors instead
> No overloads for "__init__" match the provided arguments

Actually, the only way I found it working for now, without errors, warning or type check nuisances, is also the way I think now it's supposed to be used, that is dumping the model using `model_dump(mode="json")`, which returns a dict of only JSON-serializable types.

Example:

```python
from pydantic import BaseModel, HttpUrl


class MyModel(BaseModel):
    my_url: HttpUrl


my_model = MyModel(my_url="http://www.github.com")
dump_python = my_model.model_dump()
dump_json = my_model.model_dump(mode="json")  # <<< this

print(my_model)
# my_url=Url('http://www.github.com/')
print(dump_python)
# {'my_url': Url('http://www.github.com/')}
print(dump_python["my_url"])
# http://www.github.com/
print(type(dump_python["my_url"]))
# <class 'pydantic_core._pydantic_core.Url'>
print(dump_json)
# {'my_url': 'http://www.github.com/'}
print(dump_json["my_url"])
# http://www.github.com/
print(type(dump_json["my_url"]))
# <class 'str'>
```

Disclaimer: this particular example produces a mypy warning for
> "Literal['http://www.github.com']" is incompatible with "HttpUrl"

but not creating the instance explicitly (for example receiving the model from a FastAPI route) won't display this.

**Comment by a-jean-andreasian** [comment]
Im reading this discussion, and if so, it's really easier to use just `str`, and do `if url.startswith` stuff than doing this vodoo dancing with type hinting.

Anyways `HttpUrlString = Annotated[HttpUrl, AfterValidator(str)]` this keeps warining `UserWarning: Pydantic serializer warnings:
  Expected `url` but got `str` - serialized value may not be as expected`

**Reply by vvanglro** [reply]
The same. Is there any other way besides ignoring the warning?
```py
model_dump(mode="json", warnings='none')
```

**Comment by merl-dev** [comment]
the following passes `mypy`:

```python
from typing import Annotated

from pydantic import AfterValidator, BaseModel, HttpUrl, ValidationError


class MyModel(BaseModel):
    url: Annotated[HttpUrl, AfterValidator(str)]

m1 = MyModel(url=HttpUrl('http://puny£code.com'))
assert isinstance(m1.url, str)

try:
    m2 = MyModel(url=HttpUrl('ftp://www.аррӏе.com/'))
except ValidationError as e:
    print(e)

try:
    m3 = MyModel(url=HttpUrl('htps://www/'))
except ValidationError as e:
    print(e)
```
Isn't this the expected behavior?

**Comment by patrsc** [comment]
I created a Python package [pydantic-string-url](https://pypi.org/project/pydantic-string-url/), which can be used as a replacement for Pydantic's URL types `AnyUrl`, `AnyHttpUrl`, `HttpUrl`, `AnyWebsocketUrl`, `WebsocketUrl`, `FileUrl`, and `FtpUrl`. Those types are based on `str` and can be directly passed to library packages that expect strings. They also preserve the absence of trailing slashes. Usage example:

```py
from pydantic import BaseModel, TypeAdapter, ValidationError
from pydantic_string_url import HttpUrl

class User(BaseModel):
    name: str
    age: int
    homepage: HttpUrl

user = {"name": "John Doe", "age": 33, "homepage": "https://example.com"}
invalid_user = {"name": "Alice", "age": 32, "homepage": "not a url"}

john = User.model_validate(user)
assert john.homepage == "https://example.com"  # no trailing slash was added
try:
    alice = User.model_validate(invalid_user)
except ValidationError as e:
    print(str(e))
```

[Feedback](https://github.com/patrsc/pydantic-string-url/issues) and bug reports are welcome. Maybe something similar makes it into Pydantic in the future.

**Reply by JackTheMico** [reply]
> I created a Python package [pydantic-string-url](https://pypi.org/project/pydantic-string-url/), which can be used as a replacement for Pydantic's URL types `AnyUrl`, `AnyHttpUrl`, `HttpUrl`, `AnyWebsocketUrl`, `WebsocketUrl`, `FileUrl`, and `FtpUrl`. Those types are based on `str` and can be directly passed to library packages that expect strings. They also preserve the absence of trailing slashes. Usage example:
> 
> ```python
> from pydantic import BaseModel, TypeAdapter, ValidationError
> from pydantic_string_url import HttpUrl
> 
> class User(BaseModel):
>     name: str
>     age: int
>     homepage: HttpUrl
> 
> user = {"name": "John Doe", "age": 33, "homepage": "https://example.com"}
> invalid_user = {"name": "Alice", "age": 32, "homepage": "not a url"}
> 
> john = User.model_validate(user)
> assert john.homepage == "https://example.com"  # no trailing slash was added
> try:
>     alice = User.model_validate(invalid_user)
> except ValidationError as e:
>     print(str(e))
> ```
> 
> [Feedback](https://github.com/patrsc/pydantic-string-url/issues) and bug reports are welcome. Maybe something similar makes it into Pydantic in the future.

Thank you! This works perfectly without any warnings popping up.

**Comment by Elyasomer** [comment]
The `encoded_string()` method worked for me.

```
from pydantic import BaseModel, HttpUrl

class User(BaseModel):
    name: str
    age: int
    homepage: HttpUrl
    
user = {"name": "John Doe", "age": 33, "homepage": "https://example.com"}
john = User.model_validate(user)
print(type(john.homepage.encoded_string())) # output: <class 'str'>
```
